src/hotspot/share/classfile/stringTable.hpp

```diff
@@ -56,6 +56,37 @@ class StringTable : AllStatic {
   static double get_load_factor();
   static double get_dead_factor(size_t num_dead);
 
+  // typedef enum { obj_str, unicode_str, symbol_str, utf8_str } StringType;
+  enum class StringType {
+    OopStr, UnicodeStr, SymbolStr, UTF8Str
+  };
+
+  struct StringWrapperInternal {
+    union {
+      const Handle oop_str;
+      const jchar* unicode_str;
+      const Symbol* symbol_str;
+      const char* utf8_str;
+    };
+    const StringType type;
+
+    StringWrapperInternal(const Handle oop_str)     : oop_str(oop_str),         type(StringType::OopStr)     {}
+    StringWrapperInternal(const jchar* unicode_str) : unicode_str(unicode_str), type(StringType::UnicodeStr) {}
+    StringWrapperInternal(const Symbol* symbol_str) : symbol_str(symbol_str),   type(StringType::SymbolStr)  {}
+    StringWrapperInternal(const char* utf8_str)     : utf8_str(utf8_str),       type(StringType::UTF8Str)    {}
+  };
+
+public:
+  typedef struct StringWrapperInternal StringWrapper;
+
+  static bool wrapped_string_equals(oop java_string, StringWrapper wrapped_str, int len);
+
+private:
+  static unsigned int hash_wrapped_string(StringWrapper wrapped_str, int len);
+  static const jchar *to_unicode(StringWrapper wrapped_str, int len, TRAPS);
+  static Handle to_handle(StringWrapper wrapped_str, int len, TRAPS);
+  static void print_string(StringWrapper wrapped_str, int len, TRAPS);
```

Comment:
What is `len` supposed to represent in all of these methods? The code only makes sense to me if `len` here is actually "number of unicode characters" (which need not be the same as the length of any wrapped UTF8 sequence).

Reply:
Actually that in itself is not enough to make the code correct AFAICS. Consider this example from Table 2-11, page 71, in https://www.unicode.org/versions/Unicode16.0.0/UnicodeStandard-16.0.pdf The Unicode character 00010384 in UTF-32, consists of the surrogate pair D800, DF84 in UTF-16, and the four byte sequence F0, 90, 8E, 84 in UTF-8. So if we have a java.lang.String that represents this single unicode character, the String will consist of an array of 2 char values, and the UTF8 representation would consist of 4 byte values. So if you were doing an equals comparison between the String's value array and the utf_str, what length would you pass to the equals method?

Reply:
I'd hope that the answer is 4, but I suspect that the answer is 5 (UTF-16). When the class `UTF8` talks about "unicode", it seems to be talking about whatever encoding Java's strings are, which AFAIK is basically UTF-16 with some special cases.

Reply:
See for example this snippet from `java_lang_String::create_from_str(const char* utf8_str, T/
```c++
  int length = UTF8::unicode_length(utf8_str, is_latin1, has_multibyte);
  if (!CompactStrings) {
    has_multibyte = true;
    is_latin1 = false;
  }

  Handle h_obj = basic_create(length, is_latin1, CHECK_NH);
  if (length > 0) {
    if (!has_multibyte) {
      const jbyte* src = reinterpret_cast<const jbyte*>(utf8_str);
      ArrayAccess<>::arraycopy_from_native(src, value(h_obj()), typeArrayOopDesc::element_offset<jbyte>(0), length);
    } else if (is_latin1) {
      UTF8::convert_to_unicode(utf8_str, value(h_obj())->byte_at_addr(0), length);
    } else {
      UTF8::convert_to_unicode(utf8_str, value(h_obj())->char_at_addr(0), length);
    }
  }
```

Reply:
The `len` in all these methods is the number of unicode characters, which yes, could be less than the length of the UTF8 array.

---

`UTF8::convert_to_unicode` uses `UTF8::next`, like I also do in the new UTF8 equals/hash functions. Part of `UTF8::next` looks like this: 

```c++
  switch ((ch = ptr[0]) >> 4) {
    default:
    ... work ... /* 1-byte character */
    break;

  case 0x8: case 0x9: case 0xA: case 0xB: case 0xF:
    /* Shouldn't happen. */
    break;

  case 0xC: case 0xD:
    /* 110xxxxx  10xxxxxx */
    ... work ... /* 2-byte character */
    break;

  case 0xE:
    /* 1110xxxx 10xxxxxx 10xxxxxx */
    ... work ... /* 3-byte character */
    break;
  } /* end of switch */
```

In this code, 4-byte long UTF8 characters are not converted. This leads me to believe that we do not support this range of characters. With this restriction, we also do not have 2 wide (4 byte) UTF16 characters, as 3-byte UTF8 characters fit in a single (2 byte) UTF16 unit.

So for the question, I do not know what would happen. I believe it would be undefined behaviour as this character would not be supported. Just like if we would use `UTF8::convert_to_unicode` first to then compare two UTF16 strings.

Reply:
@caspernorrbin - Aarggh! Sorry I picked a bad example - the JVM uses modified-UTF8 which does not support 4-byte UTF8 encodings. Any UTF16 surrogate pair would be encoded as two three-byte modifed-UTF8 sequences.

> The len in all these methods is the number of unicode characters

Are you sure you meant "characters" there?

Let me try to back up a bit and go to the code that I have a problem with:
```
bool java_lang_String::equals(oop java_string, const char* utf8_string, int num_unicode_points) {
  assert(java_string->klass() == vmClasses::String_klass(),
         "must be java_string");
  typeArrayOop value = java_lang_String::value_no_keepalive(java_string);
  int length = java_lang_String::length(java_string, value);
  if (length != num_unicode_points) {
    return false;
  }
```
The incoming String represents a single unicode character encoded as a surrogate pair - so `length == 2`. Each element in the surrogate pairs encodes as a 3-byte modified-UTF8 sequence, so the length of `utf8_string` is 6. For the length check to pass the incoming value of `num_unicode_points` has to be 2. So for this to succeed the caller always has to pass in the number of UTF16 elements that comprise the `java_string`. As Johan noted in the utf8.cpp file when it refers to "unicode" it really means UTF16, so in that sense `num_unicode_points` is not incorrect.  But it means those `len` parameters should also be named the same way.

Reply:
You're right, I meant the number of code points..

The `num_unicode_points` in this case would come from `UTF8::unicode_length`, which would give 2 from the 6-byte UTF8 string in your example. But I agree that it is somewhat unclear, especially if someone were to use this function in the future.

I have two ideas how this could be improved:
I could either add clarifying comments and change naming across the different functions to be clearer on how to use and call correctly. Or I could instead change it so that the function takes the UTF8-length instead of the UTF16-length. That way you always pass the array length to the function.

Reply:
I just want consistency and clarity in the header prototypes, for me either option is fine.